File: apps/frontend/lib/src/features/announcement/data/dto/create_announcement_dto.dart
class CreateAnnouncementDto {
  const CreateAnnouncementDto({
    required this.title,
    required this.message,
    required this.category,
    required this.priority,
    required this.target_audience,
    this.target_roles,
    this.scheduled_at,
    this.expires_at,
    this.publish_immediately,
  });

  final String title;
  final String message;
  final String category;
  final String priority;
  final String target_audience;
  final List<String>? target_roles;
  final String? scheduled_at;
  final String? expires_at;
  final bool? publish_immediately;

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'message': message,
      'category': category,
      'priority': priority,
      'targetAudience': target_audience,
      if (target_roles != null && target_roles!.isNotEmpty)
        'targetRoles': target_roles,
      if (scheduled_at != null) 'scheduledAt': scheduled_at,
      if (expires_at != null) 'expiresAt': expires_at,
      if (publish_immediately != null)
        'publishImmediately': publish_immediately,
    };
  }
}

File: apps/backend/src/modules/announcement/dto/create-announcement.dto.ts
import {
  IsString,
  IsNotEmpty,
  IsOptional,
  IsEnum,
  IsArray,
  IsBoolean,
  IsDateString,
  MaxLength,
  ValidateIf,
} from 'class-validator';
import { Expose } from 'class-transformer';
import { AnnouncementCategory } from '../enums/announcement-category.enum';
import { AnnouncementPriority } from '../enums/announcement-priority.enum';
import { AnnouncementTargetAudience } from '../enums/announcement-target-audience.enum';

export class CreateAnnouncementDto {
  @Expose()
  @IsString()
  @IsNotEmpty()
  @MaxLength(255)
  title!: string;

  @Expose()
  @IsString()
  @IsNotEmpty()
  message!: string;

  @Expose()
  @IsEnum(AnnouncementCategory)
  @IsNotEmpty()
  category!: AnnouncementCategory;

  @Expose()
  @IsEnum(AnnouncementPriority)
  @IsNotEmpty()
  priority!: AnnouncementPriority;

  @Expose()
  @IsEnum(AnnouncementTargetAudience)
  @IsNotEmpty()
  targetAudience!: AnnouncementTargetAudience;

  @Expose()
  @IsArray()
  @IsString({ each: true })
  @ValidateIf((o) => o.targetAudience === AnnouncementTargetAudience.ROLES)
  @IsNotEmpty()
  targetRoles?: string[];

  @Expose()
  @IsDateString()
  @IsOptional()
  scheduledAt?: string;

  @Expose()
  @IsDateString()
  @IsOptional()
  expiresAt?: string;

  @Expose()
  @IsBoolean()
  @IsOptional()
  publishImmediately?: boolean;
}

